lua/lspconfig/configs.lua

```diff
@@ -87,13 +87,31 @@ function configs.__newindex(t, config_name, config_def)
         root_dir = get_root_dir(util.path.sanitize(bufname), bufnr)
       end
 
-      if root_dir then
-        api.nvim_command(
+      if config.single_file_support then
```

Comment:
This doesn't seem correct? should we not prefer the "true" root directory resolved from the root function instead of using the pseudo root derived from the parent directory of the buffer?

Reply:
So, here the diff doesn't really convey what I did. I swapped the two branches of the if statement. The problem with the previous version is that most of the time, even when `single_file_support` is set, a valid `root_dir` is computed, meaning that the `config.single_file_support` branch of the if is not executed. 

Moreover the `root_dir` branch sets autocommands for any file in the same `root_dir`.
I don't know if this was intentional, but if I activate `single_file_support` I don't want the server to start for every file in the same root, I just want it for a "single file".

Reply:
In other words: `single_file_support` should take precedence when is set, otherwise we fall back to standard behaviour (compute root and set the server to start/attach for any file in the root).

Reply:
> In other words: single_file_support should take precedence when is set

I think this is a misunderstanding of the point of single_file_support, the point of single_file_support is to be a fallback *after* we try resolving the project level directory pattern with root_dir, it results in degraded server performance if we send a null root directory.

Reply:
Mmm maybe I did misunderstand the snippet of code.
But what is the usecase of SFS option then? Let me give you an example:
I use lTeX as a per-file toggle syntax/grammar checker (the toggling does other per-buffer things on top of starting the server). My config for the server has SFS set to true and autostart to false.
The current version of the plugin, when I turn on the server, manages to find a valid root, starts the server and sets the autocmds. The snippet of code that is executed is *the same* whether I (un)set the SFS option.  Now, due to the autocommands being set, when I open a new file the server attaches to these files as well (what? I asked for SFS and no autostart, why the server is starting on its own now?).

At this point maybe the code should be something like
```lua
if SFS then
  -- SFS branch
elseif valid_root then
  -- proceed in standard way and set autocommands
else
  -- compute parent dir as root and set autocommands
end
```

(Setting root and autocommands can be split in different if statements. This is just to mimic the current workflow.)

Reply:
> But what is the usecase of SFS option then

I suggest reading `:help lspconfig-single-file-suppor`

Reply:
Ugh, sorry, the naming of the feature is not very intuitive, but the docs clearly state what it is... my bad!. I can revert that change. 

I still think the other changes are valid, tho. And at this point I would suggest  not to set autocmds when `autostart` is unset.

Reply:
> sorry, the naming of the feature is not very intuitive

How so? This allows launching language servers on single, isolated files outside of a project. I'm happy to rename it, but I think it's intuitive (this is the first time this has come up since I added it last year). 

> I still think the other changes are valid, tho. And at this point I would suggest not to set autocmds when autostart is unset.

Sure, please push the latest changes and I will take a look.

Reply:
I don't really have a better name for the option, but I was expecting that, by setting it to true I would get be able to launch the single file, regardless of where the file is. But this is not the case, you get this behaviour only when a better root is not found.

To be fair, my misunderstanding might be due to the fact that I was trying to define a custom behaviour for lTeX which is technically a LS but with a different usecase.

That said, I pushed a new version.

> And at this point I would suggest not to set autocmds when autostart is unset.

This was stupid, ignore it :) We still want the autocommands when we manually start a server. We just need to make sure to remove them when we stop the server.

Reply:
>  I was expecting that, by setting it to true I would get be able to launch the single file, regardless of where the file is.

I mean, that is the case, it just attempts to resolve the root directory first and then falls back, but you will *always* launch the server where it does attach to the file. If you want to *only* use the single file, I don't see why you wouldn't just directly calls `lsp.start_client` and bypass lspconfig entirely.

Reply:
Yeah, I can do that. I simply thought I could do this with LspConfig, so I didn't bother using the low-level function.